### PR TITLE
MM-46392 - create channels with a board

### PR DESCRIPTION
**Pull request by pvev**
#### Summary
This PR adds the logic to be able to create a new channel along with a new board associated to the created channel. 

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-46392

#### Related Pull Requests
n/a

#### Screenshots

https://user-images.githubusercontent.com/10082627/189082208-32eed793-1c93-46cf-ab86-c443206e08fc.mp4

https://user-images.githubusercontent.com/10082627/189082282-8ca10e91-48c8-4aa7-969b-4cf1841c681c.mp4

#### Release Note

```release-note
ability to create a new channel along with a new board associated to the created channel
```

### CODE REVIEW

**Comment by pvev** [comment]
@julmondragon @JulienTant could you please help me with an early feedback for this PR? It is still missing e2e tests and an scenario for showing the tourtip in the top app bar, but it will really helpful to receive early feedback on this one. Thanks!

File: components/new_channel_modal/new_channel_modal.tsx [hunk 1]
```diff
@@ -219,6 +280,41 @@ const NewChannelModal = () => {
...
+            <Tooltip id='new-board-tooltip'>
+                <>
+                    <div style={{fontWeight: 'bold'}}>
+                        <FormattedMessage
+                            id={'channel_modal.create_board.tooltip'}
+                            defaultMessage={'Manage your task with a board'}
+                        />
+                    </div>
+                    <FormattedMessage
+                        id={'channel_modal.create_board.tooltip'}
```
Comment (JulienTant) [inline components/new_channel_modal/new_channel_modal.tsx:300]: Those `id` are the same here

File: components/new_channel_modal/new_channel_modal.tsx [hunk 2]
```diff
@@ -116,12 +135,54 @@ const NewChannelModal = () => {
...
+            };
+            const {data: patchedBoard} = await dispatch(attachBoardToChannel(newBoard.id, boardPatch));
+
+            if (patchedBoard.channelId === channelId) {
+                const boardUrl = `${getSiteURL()}/boards/team/${currentTeamId}/${newBoard.id}`;
+
+                const newBoardMsg = formatMessage(
+                    {
+                        id: 'channel_modal.board.newBoardCreated2',
+                        defaultMessage: '**{user}** created the board [{board}]({boardUrl}) and linked it to this channel.',
```
Comment (JulienTant) [inline components/new_channel_modal/new_channel_modal.tsx:168]: I don't know how it happens but there's a weird double space in the app coming. Maybe adding a trim to the board title would do the trick?
image

**Review by JulienTant — CHANGES_REQUESTED** [review]
The feature works great, just found a thing that needs fixing (same translation ID) and one minor issue (double space).
Good job @pvev

File: packages/types/src/posts.ts [hunk 3]
```diff
@@ -27,6 +27,7 @@ export type PostType = 'system_add_remove' |
 'system_remove_from_channel' |
 'system_combined_user_activity' |
 'system_fake_parent_deleted' |
+'system_generic' | 'system_' |
```
Comment (julmondragon) [inline packages/types/src/posts.ts:30]: What's `system_ ` used for?

**Review by julmondragon — COMMENTED** [review]
Great work @pvev !
Just some questions and nits after the early review.

File: packages/types/src/boards.ts [hunk 4]
```diff
@@ -11,3 +11,67 @@ export declare type BoardsUsageResponse = {
...
+    type?: BoardTypes;
+    minimumRole?: string;
+    title?: string;
+    description?: string;
+    icon?: string;
+    showDescription?: boolean;
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    updatedProperties?: Record<string, any>;
+    deletedProperties?: string[];
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
```
[the remark this answers is not in the document]
Comment (pvev) [reply packages/types/src/boards.ts:70]: not necessary, good catch!

File: packages/types/src/boards.ts [hunk 5]
```diff
@@ -11,3 +11,67 @@ export declare type BoardsUsageResponse = {
     card_limit_timestamp: number;
 };
 
+const BoardTypeOpen = 'O';
+const BoardTypePrivate = 'P';
+const boardTypes = [BoardTypeOpen, BoardTypePrivate];
+type BoardTypes = typeof boardTypes[number];
+
+type PropertyTypeEnum = 'text' | 'number' | 'select' | 'multiSelect' | 'date' | 'person' | 'file' | 'checkbox' | 'url' | 'email' | 'phone' | 'createdTime' | 'createdBy' | 'updatedTime' | 'updatedBy' | 'unknown'
```
[the remark this answers is not in the document]
Comment (pvev) [reply packages/types/src/boards.ts:19]: ok!

File: packages/mattermost-redux/src/actions/boards.ts [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+        const state = getState();
+        const currentUserId = getCurrentUserId(state);
+        const preference: PreferenceType = {
+            user_id: currentUserId,
+            category: Preferences.APP_BAR,
+            name: Preferences.NEW_CHANNEL_WITH_BOARD_TOUR_SHOWED,
+            value: JSON.stringify(initializationState),
+        };
+        await dispatch(savePreferences(currentUserId, [preference]));
+        return {data: true};
```
[the remark this answers is not in the document]
Comment (pvev) [reply packages/mattermost-redux/src/actions/boards.ts:82]: no need for it, nothing will be done with that value, only returning this for satisfying TS in terms of the ActionFunc return value.

File: components/new_channel_modal/new_channel_modal.tsx [hunk 7]
```diff
@@ -219,6 +280,41 @@ const NewChannelModal = () => {
 
     const canCreate = displayName && !displayNameError && url && !urlError && type && !purposeError && !serverError;
 
+    const addBoardToggle = async () => {
+        setAddBoard(!addBoard);
+        const {data: boardTemplates} = await dispatch(getBoardsTemplates());
+        setBoardTemplates(boardTemplates);
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:286]: not at all, good catch, will add a validation for it.

File: components/new_channel_modal/new_channel_modal.tsx [hunk 8]
```diff
@@ -116,12 +135,54 @@ const NewChannelModal = () => {
             }
 
             handleOnModalCancel();
+
+            // If template selected, create a new board from this template
+            if (addBoard && selectedBoardTemplate) {
+                addBoardToChannel(newChannel.id);
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:141]: yes, it is better to make this process synchronous.

[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:141]: will add some error thrown and log it to the console, but the idea is not block the channel process creation.

File: components/new_channel_modal/new_channel_modal.tsx [hunk 9]
```diff
@@ -116,12 +135,54 @@ const NewChannelModal = () => {
...
+                    {
+                        id: 'channel_modal.board.newBoardCreated2',
+                        defaultMessage: '**{user}** created the board [{board}]({boardUrl}) and linked it to this channel.',
+                    },
+                    {
+                        user: user.username,
+                        board: newBoard.title,
+                        boardUrl,
+                    },
+                );
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:175]: For this kind of ephemeral post messages is the way it works, the process to change it is out of scope and probably not wanted. Using Format function breaks it, so markdown text is used.

File: components/new_channel_modal/new_channel_modal.tsx [hunk 10]
```diff
@@ -219,6 +280,41 @@ const NewChannelModal = () => {
 
     const canCreate = displayName && !displayNameError && url && !urlError && type && !purposeError && !serverError;
 
+    const addBoardToggle = async () => {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:283]: ok!

File: components/new_channel_modal/new_channel_modal.tsx [hunk 11]
```diff
@@ -301,6 +397,58 @@ const NewChannelModal = () => {
...
+                        <label>
+                            <input
+                                type='checkbox'
+                                onChange={addBoardToggle}
+                                checked={addBoard}
+                            />
+                            {formatMessage({id: 'channel_modal.create_board.title', defaultMessage: 'Create a board for this channel'})}
+                        </label>
+                        {newBoardInfoIcon()}
+                    </div>
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:410]: no need for it, any user can create boards.

File: components/new_channel_modal/new_channel_modal.tsx [hunk 12]
```diff
@@ -301,6 +397,58 @@ const NewChannelModal = () => {
...
+                            >
+                                <Input
+                                    type='email'
+                                    autoComplete='off'
+                                    value={selectedBoardTemplate?.title ? `${selectedBoardTemplate.icon} ${selectedBoardTemplate.title}` : ''}
+                                    name='select-board-template'
+                                    containerClassName='select-board-template-container'
+                                    inputClassName='select-board-template-input'
+                                    label={formatMessage({id: 'channel_modal.create_board.select_template', defaultMessage: 'Select a template'})}
+                                    placeholder={formatMessage({id: 'channel_modal.create_board.select_template_placeholder', defaultMessage: 'Select a template'})}
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:425]: yes according to design.

File: components/new_channel_modal/new_channel_modal.tsx [hunk 13]
```diff
@@ -301,6 +397,58 @@ const NewChannelModal = () => {
...
+                                />
+                                <Menu
+                                    openLeft={true}
+                                    openUp={false}
+                                    id='SelectBoardTemplate'
+                                    className='SelectTemplateMenu'
+                                    ariaLabel={intl.formatMessage({id: 'channel_modal.create_board.select_template', defaultMessage: 'Select board template'})}
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:432]: ok! good catch!

File: components/new_channel_modal/new_channel_modal.tsx [hunk 14]
```diff
@@ -301,6 +397,58 @@ const NewChannelModal = () => {
...
+                                >
+                                    <Menu.Group>
+                                        {boardTemplates?.map((boardTemplate: BoardTemplate) => {
+                                            return (
+                                                <Menu.ItemAction
+                                                    id={boardTemplate.title.replace(' ', '_')}
+                                                    onClick={() => setSelectedBoardTemplate(boardTemplate)}
+                                                    icon={boardTemplate.icon || <i className='icon icon-product-boards'/>}
+                                                    text={boardTemplate.title}
+                                                    extraText={boardTemplate.description ? `${boardTemplate.description.substring(0, 70)} ...` : ''}
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:442]: ok! will check for length of the string and if required will use ellipsis char.

File: components/new_channel_modal/new_channel_modal.tsx [hunk 15]
```diff
@@ -301,6 +397,58 @@ const NewChannelModal = () => {
...
+                                />
+                                <Menu
+                                    openLeft={true}
+                                    openUp={false}
+                                    id='SelectBoardTemplate'
+                                    className='SelectTemplateMenu'
+                                    ariaLabel={intl.formatMessage({id: 'channel_modal.create_board.select_template', defaultMessage: 'Select board template'})}
+                                >
+                                    <Menu.Group>
+                                        {boardTemplates?.map((boardTemplate: BoardTemplate) => {
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:435]: won't be big nor contain lots of data. Doing memoization/useCallback is also expensive in terms of performance. Most of the times is cheaper to let React to re-render

File: components/channel_header/components/header_icon_wrapper.tsx [hunk 16]
```diff
@@ -160,20 +163,34 @@ const HeaderIconWrapper: React.FC<Props> = (props: Props) => {
                         {iconComponent}
                     </button>
                 </OverlayTrigger>
+                {pluginId && pluginId === 'focalboard' &&
+                    <NewChannelWithBoardTourTip
+                        pulsatingDotPlacement={'start'}
+                        pulsatingDotTranslate={{x: 0, y: -22}}
+                    />
+                }
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/channel_header/components/header_icon_wrapper.tsx:171]: ok

File: components/channel_header/components/header_icon_wrapper.tsx [hunk 17]
```diff
@@ -160,20 +163,34 @@ const HeaderIconWrapper: React.FC<Props> = (props: Props) => {
                         {iconComponent}
                     </button>
                 </OverlayTrigger>
+                {pluginId && pluginId === 'focalboard' &&
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/channel_header/components/header_icon_wrapper.tsx:166]: it is indeed specific, I agree with your approach too, but 4/5 keeping it this way following the way it is done in all the other places where we show the tourtips.

File: components/app_bar/app_bar.tsx [hunk 18]
```diff
@@ -42,12 +42,14 @@ export default function AppBar() {
...
-                />
-            ))}
+            {coreProducts.map((product) => {
+                return (
+                    <AppBarPluginComponent
+                        key={product.id}
+                        component={product}
+                    />
+                );
+            })}
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/app_bar/app_bar.tsx:52]: no reason, was just a leftover of initial implementation code.

File: components/app_bar/new_channel_with_board_tour_tip.tsx [hunk 19]
```diff
@@ -0,0 +1,91 @@
...
+import {setNewChannelWithBoardPreference} from 'mattermost-redux/actions/boards';
+import {Preferences} from 'mattermost-redux/constants';
+
+import {showNewChannelWithBoardPulsatingDot} from 'selectors/plugins';
+
+import TourTip, {useMeasurePunchouts} from 'components/widgets/tour_tip';
+import FormattedMarkdownMessage from 'components/formatted_markdown_message.jsx';
+
+type Props = {
+    pulsatingDotPlacement?: string;
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/app_bar/new_channel_with_board_tour_tip.tsx:17]: sure!

File: components/app_bar/new_channel_with_board_tour_tip.tsx [hunk 20]
```diff
@@ -0,0 +1,91 @@
...
+
+    const title = (
+        <FormattedMessage
+            id='newChannelWithBoard.tutorialTip.title'
+            defaultMessage='Access linked boards from the App Bar'
+        />
+    );
+
+    const screen = (
+        <FormattedMarkdownMessage
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/app_bar/new_channel_with_board_tour_tip.tsx:36]: ok!

File: components/app_bar/app_bar.scss [hunk 21]
```diff
@@ -136,3 +136,30 @@ $channel-view-padding: 63px;
 .app-bar::-webkit-scrollbar {
     display: none;
 }
+
+// this style is defined outside of the block scope because is intended to affect the tippy element
```
[the remark this answers is not in the document]
Comment (pvev) [reply components/app_bar/app_bar.scss:140]: ok!

File: packages/types/src/posts.ts (hunk 3 above)
Comment (pvev) [reply packages/types/src/posts.ts:30]: is a left-over, will remove it. Thanks

File: components/new_channel_modal/new_channel_modal.tsx (hunk 1 above)
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:300]: nice catch! thanks!

File: components/new_channel_modal/new_channel_modal.tsx (hunk 2 above)
Comment (pvev) [reply components/new_channel_modal/new_channel_modal.tsx:168]: ok, will check that. Thanks!

**Comment by annelieseklein** [comment]
Looks great!

Screen Shot 2022-09-15 at 10 35 33 AM
1.  Select template field should have a dropdown trailing icon
2. For the template dropdown, can we reduce it to showing just 3 templates and not 4? feels a bit long with 4.
3. For the tooltip on the ? the first line should be in the header style font and all of the text should be left aligned
4. Clicking on the board does not take the user to the boards itself, it takes me to the template picker in Boards. Is it possible for it to just take me to the board I created?

**Comment by pvev** [comment]
> looking really great @pvev !

Captura de Pantalla 2022-09-28 a las 16 04 17

https://user-images.githubusercontent.com/10082627/192833878-065bb991-96f2-4e3a-baa7-6bec65fdf95d.mp4

@annelieseklein ^^

**Comment by annelieseklein** [comment]
Screen Shot 2022-09-28 at 9 23 08 AM

when i go to edit the URL and hit the "Enter" key on my keyboard, instead of selecting "Done" for editing the URL it goes ahead and creates the channel. When editing the URL, can we have clicking enter means clicking "done" and closes the editing and not create the channel @pvev

**Comment by pvev** [comment]
> Screen Shot 2022-09-28 at 9 23 08 AM
> 
> when i go to edit the URL and hit the "Enter" key on my keyboard, instead of selecting "Done" for editing the URL it goes ahead and creates the channel and closes the modal. I did not want to create the channel just yet, I just wanted to finish editing the URL. When editing the URL, can we have clicking enter means clicking "done" and closes the editing and not create the channel @pvev

ok, I see,  we can file a new bug, so we keep the work done here in this PR in the scope of the ticket.

**Comment by annelieseklein** [comment]
Current implementation: Screen Shot 2022-09-28 at 9 27 44 AM


Design implementation: 
Screen Shot 2022-09-28 at 9 27 34 AM


1. there are color differences between the two checkboxes --- (it should be 24% center channel text)
2. There are color differences between the two texts --- (it should be center channel text)
3. there are differences in the size of the icon --- (it should be 18px)
@pvev

File: components/new_channel_modal/new_channel_modal.tsx [hunk 22]
```diff
@@ -109,19 +135,89 @@ const NewChannelModal = () => {
...
             }
 
             handleOnModalCancel();
+
+            // If template selected, create a new board from this template
+            if (addBoard && selectedBoardTemplate) {
+                try {
+                    addBoardToChannel(newChannel.id);
+                } catch (e: any) {
+                    console.log(e.message); // eslint-disable-line
```
Comment (julmondragon) [inline components/new_channel_modal/new_channel_modal.tsx:150]: Is this actually wanted for prod?

**Review by julmondragon — APPROVED** [review]
Looks good! Nice work @pvev

**Review by annelieseklein — APPROVED** [review]
looks great! great work pablo! :)

**Review by JulienTant — APPROVED** [review]
Code looks good, feature works great! 🚢

**Comment by marianunez** [comment]
@pvev @stevemudie can we prioritize this PR ? Anything pending to move forward on this?

**Comment by pvev** [comment]
> @pvev @stevemudie can we prioritize this PR ? Anything pending to move forward on this?

@stevemudie let me know if you need any help at validating this PR

**Comment by pvev** [comment]
> @pvev Sorry for the delay in this. It looks good but should it show "Welcome to boards!" in the dropdown when selecting a template? Choosing that as the 'template' fails to create a board.
> 
> Screenshot 2022-10-17 at 12 59 09 PM
Well, it is a template, but I don't see it failing creating the template.

**Comment by pvev** [comment]
> > @pvev Sorry for the delay in this. It looks good but should it show "Welcome to boards!" in the dropdown when selecting a template? Choosing that as the 'template' fails to create a board.
> > Screenshot 2022-10-17 at 12 59 09 PM
> > Well, it is a template, but I don't see it failing creating the template @stevemudie ^^.

**Comment by stevemudie** [comment]
🤔 
I guess I took it that it wasn't working because it fails to appear on the RHS like the other 2 do when the channel & board is created.

It's there though if I go to boards so all set then as is @pvev?

Screenshot 2022-10-17 at 2 03 35 PM

**Review by stevemudie — APPROVED** [review]
All set @pvev!

**Comment by hmhealey** [comment]
FYI, my team started a discussion in the Growth Team channel in community about this since there's some issues with how it was done, particularly around accessing Boards functionality from within the Core App/Channels. We can keep the discussion on that to the thread, but I wanted to mention it here for increased visibility

https://community.mattermost.com/private-core/pl/gb5b639n3t84xrwkbrey9o7x7c

**Pull request merged**